File: Sample/java/src/main/java/net/nend/sample/java/nativead/NativeRecyclerActivity.java
package net.nend.sample.java.nativead;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import net.nend.android.NendAdNative;
import net.nend.android.NendAdNativeClient;
import net.nend.android.NendAdNativeListener;
import net.nend.android.NendAdNativeViewBinder;
import net.nend.sample.java.R;
import net.nend.sample.java.nativeadvideo.utilities.MyNendAdViewHolder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;

public class NativeRecyclerActivity extends AppCompatActivity {

    private final int NORMAL = 0;
    private final int AD = 1;
    private final String TAG = getClass().getSimpleName();
    // 広告を表示したポジションの一覧
    private List<Integer> mPositionList = new ArrayList<>();
    // 表示したポジションと広告を紐付けて保持
    private HashMap<Integer, NendAdNative> mLoadedAd = new HashMap<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.native_recycler);

        ArrayList<String> list = new ArrayList<>();
        for (int i = 1; i < 100; i++) {
            list.add("item" + i);
        }

        RecyclerView recyclerView = findViewById(R.id.recycler);
        assert recyclerView != null;
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        recyclerView.setAdapter(new NativeRecyclerAdapter(this, list));
    }

    class NativeRecyclerAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

        private LayoutInflater mLayoutInflater;
        private List<String> mList;
        private NendAdNativeClient mClient;
        private NendAdNativeViewBinder mBinder;

        public NativeRecyclerAdapter(Context context, List<String> list) {
            super();

            mLayoutInflater = LayoutInflater.from(context);
            mList = list;

            mBinder = new NendAdNativeViewBinder.Builder()
                    .adImageId(R.id.ad_image)
                    .titleId(R.id.ad_title)
                    .promotionNameId(R.id.ad_promotion_name)
                    .prId(R.id.ad_pr, NendAdNative.AdvertisingExplicitly.PR)
                    .build();
            mClient = new NendAdNativeClient(context, 485516, "16cb170982088d81712e63087061378c71e8aa5c");
        }

        @Override
        public int getItemCount() {
            return mList.size();
        }

        @Override
        public int getItemViewType(int position) {
            return (position != 0 && position % 5 == 0) ? AD : NORMAL;
        }

        @NonNull
        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
            View view;
            RecyclerView.ViewHolder viewHolder = null;
            switch (viewType) {
                case NORMAL:
                    view = mLayoutInflater.inflate(R.layout.native_list_row, viewGroup, false);
                    viewHolder = new ViewHolder(view);
                    break;
                case AD:
                    view = mLayoutInflater.inflate(R.layout.native_ad_left_row, viewGroup, false);
                    viewHolder = new MyNendAdViewHolder(view, mBinder);
                    break;
            }
            assert viewHolder != null;
            return viewHolder;
        }

        @Override
        @SuppressLint("RecyclerView")
        public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder viewHolder, final int position) {
            switch (getItemViewType(position)) {
                case NORMAL:
                    ((ViewHolder) viewHolder).textView.setText(mList.get(position));
                    ((ViewHolder) viewHolder).imageView.setBackgroundColor(Color.LTGRAY);
                    break;
                case AD:
                    if (mLoadedAd.containsKey(position)) {
                        MyNendAdViewHolder holder = (MyNendAdViewHolder) viewHolder;
                        Objects.requireNonNull(mLoadedAd.get(position)).intoView(holder.itemView, holder.normalBinder);
                        break;
                    }

                    mClient.loadAd(new NendAdNativeClient.Callback() {
                        @Override
                        public void onSuccess(final NendAdNative nendAdNative) {
                            Log.i(TAG, "広告取得成功");
                            mLoadedAd.put(position, nendAdNative);
                            mPositionList.add(position);
                            viewHolder.setIsRecyclable(false);
                            MyNendAdViewHolder holder = (MyNendAdViewHolder) viewHolder;
                            final NendAdNative adNative = mLoadedAd.get(position);
                            if (adNative != null) {
                                adNative.intoView(holder.itemView, mBinder);
                                adNative.setNendAdNativeListener(new NendAdNativeListener() {
                                    @Override
                                    public void onImpression(@NonNull NendAdNative nendAdNative) {
                                        Log.i(TAG, "onImpression");
                                    }

                                    @Override
                                    public void onClickAd(@NonNull NendAdNative nendAdNative) {
                                        Log.i(TAG, "onClickAd");
                                    }

                                    @Override
                                    public void onClickInformation(@NonNull NendAdNative nendAdNative) {
                                        Log.i(TAG, "onClickInformation");
                                    }
                                });
                            }
                        }

                        @Override
                        public void onFailure(NendAdNativeClient.NendError nendError) {
                            Log.i(TAG, "広告取得失敗 " + nendError.getMessage());
                            // すでに取得済みの広告をランダムで表示
                            if (!mLoadedAd.isEmpty()) {
                                Collections.shuffle(mPositionList);
                                MyNendAdViewHolder holder = (MyNendAdViewHolder) viewHolder;
                                Objects.requireNonNull(mLoadedAd.get(mPositionList.get(0))).intoView(holder.itemView, holder.normalBinder);
                            }
                        }
                    });
                    break;
            }
        }

        class ViewHolder extends RecyclerView.ViewHolder {

            TextView textView;
            ImageView imageView;

            public ViewHolder(View v) {
                super(v);
                textView = v.findViewById(R.id.title);
                imageView = v.findViewById(R.id.thumbnail);
            }
        }
    }
}

File: Sample/kotlin/src/main/java/net/nend/sample/kotlin/nativead/NativeRecyclerActivity.kt
package net.nend.sample.kotlin.nativead

import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import net.nend.android.NendAdNative
import net.nend.android.NendAdNativeClient
import net.nend.android.NendAdNativeListener
import net.nend.android.NendAdNativeViewBinder
import net.nend.sample.kotlin.R
import net.nend.sample.kotlin.databinding.NativeListRowBinding
import net.nend.sample.kotlin.databinding.NativeRecyclerBinding
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_API_KEY_SMALL_SQUARE
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_LOG_TAG
import net.nend.sample.kotlin.nativead.NativeSampleActivity.Companion.NATIVE_SPOT_ID_SMALL_SQUARE
import net.nend.sample.kotlin.nativeadvideo.utilities.MyNendAdViewHolder

class NativeRecyclerActivity : AppCompatActivity() {

    private lateinit var binding: NativeRecyclerBinding

    private enum class ViewType(val id: Int) {
        NORMAL(0),
        AD(1);

        companion object {
            fun from(id: Int): ViewType? = when (id) {
                0 -> NORMAL
                1 -> AD
                else -> null
            }
        }
    }

    // 広告を表示したポジションの一覧
    private val positionList = mutableListOf<Int>()

    // 表示したポジションと広告を紐付けて保持
    private val loadedAd = mutableMapOf<Int, NendAdNative>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = NativeRecyclerBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val list = (1..99).map { i -> "item$i" }
        binding.run {
            recycler.run {
                layoutManager = LinearLayoutManager(this@NativeRecyclerActivity)
                adapter = NativeRecyclerAdapter(list)
            }
        }
    }

    internal inner class NativeRecyclerAdapter(
        private val list: List<String>
    ) : ListAdapter<String, RecyclerView.ViewHolder>(DiffCallback) {
        private lateinit var client: NendAdNativeClient
        private val binder: NendAdNativeViewBinder = NendAdNativeViewBinder.Builder()
            .adImageId(R.id.ad_image)
            .titleId(R.id.ad_title)
            .promotionNameId(R.id.ad_promotion_name)
            .prId(R.id.ad_pr, NendAdNative.AdvertisingExplicitly.PR)
            .build()

        override fun getItemCount() = list.size

        override fun getItemViewType(position: Int) = if (position != 0 && position % 5 == 0) {
            ViewType.AD.id
        } else {
            ViewType.NORMAL.id
        }

        override fun onCreateViewHolder(
            viewGroup: ViewGroup,
            viewType: Int
        ): RecyclerView.ViewHolder {
            val layoutInflater: LayoutInflater = LayoutInflater.from(viewGroup.context)
            client = NendAdNativeClient(
                viewGroup.context,
                NATIVE_SPOT_ID_SMALL_SQUARE,
                NATIVE_API_KEY_SMALL_SQUARE
            )

            val viewHolder = when (ViewType.from(viewType)) {
                ViewType.AD -> {
                    MyNendAdViewHolder(
                        layoutInflater.inflate(
                            R.layout.native_ad_left_row,
                            viewGroup,
                            false
                        ),
                        binder
                    )
                }
                ViewType.NORMAL -> {
                    val binding = NativeListRowBinding.inflate(
                        layoutInflater,
                        viewGroup,
                        false
                    )
                    ViewHolder(binding)
                }
                else -> null
            } ?: throw Exception("Unexpected ViewType")

            return viewHolder
        }

        override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
            when (viewHolder) {
                is ViewHolder -> {
                    viewHolder.bind(viewHolder.layoutPosition)
                }
                is MyNendAdViewHolder -> {
                    if (loadedAd.containsKey(viewHolder.layoutPosition)) {
                        loadedAd[viewHolder.layoutPosition]?.run {
                            setNativeAdViewHolder(this, viewHolder)
                        }
                    } else {
                        client.loadAd(object : NendAdNativeClient.Callback {
                            override fun onSuccess(nendAdNative: NendAdNative) {
                                Log.i(NATIVE_LOG_TAG, "広告取得成功")
                                positionList.add(viewHolder.layoutPosition)
                                viewHolder.setIsRecyclable(false)
                                nendAdNative.run {
                                    setNativeAdViewHolder(nendAdNative, viewHolder)
                                    setNendAdNativeListener(object : NendAdNativeListener {
                                        override fun onImpression(ad: NendAdNative) {
                                            Log.i(NATIVE_LOG_TAG, "onImpression")
                                        }

                                        override fun onClickAd(ad: NendAdNative) {
                                            Log.i(NATIVE_LOG_TAG, "onClickAd")
                                        }

                                        override fun onClickInformation(ad: NendAdNative) {
                                            Log.i(NATIVE_LOG_TAG, "onClickInformation")
                                        }
                                    })
                                    loadedAd[viewHolder.layoutPosition] = this
                                }
                            }

                            override fun onFailure(nendError: NendAdNativeClient.NendError) {
                                Log.i(NATIVE_LOG_TAG, "広告取得失敗: ${nendError.message} ")
                                // すでに取得済みの広告をランダムで表示
                                if (loadedAd.isNotEmpty()) {
                                    positionList.shuffle()
                                    loadedAd[positionList[0]]?.run {
                                        setNativeAdViewHolder(this, viewHolder)
                                    }
                                }
                            }
                        })
                    }
                }
            }
        }

        private fun setNativeAdViewHolder(
            nativeAd: NendAdNative,
            holder: MyNendAdViewHolder,
        ) {
            nativeAd.intoView(
                holder.myItemView,
                holder.normalBinder,
            )
        }

        internal inner class ViewHolder(
            private val binding: NativeListRowBinding,
        ) : RecyclerView.ViewHolder(binding.root) {

            fun bind(position: Int) {
                binding.run {
                    title.text = list[position]
                    thumbnail.setBackgroundColor(Color.LTGRAY)

                    executePendingBindings()
                }
            }
        }
    }

    private object DiffCallback : DiffUtil.ItemCallback<String>() {
        override fun areItemsTheSame(oldItem: String, newItem: String): Boolean {
            return oldItem == newItem
        }

        override fun areContentsTheSame(oldItem: String, newItem: String): Boolean {
            return oldItem == newItem
        }
    }
}